[PX4: REJECTED] stm32f7 dma capable hack for SPI DMA usage

dagar opened the pull request:
This is a temporary workaround to enable SPI DMA on STM32F7. The stm32_dmacapable check typically fails in SPI dma usage because the buffer end (length) is unlikely to be dcached.

In PX4 usage the only other caller of stm32_dmacapable is sdio, where the memory is coming from the DMA allocator we supply. As a result I believe this is an acceptable short term solution until we have time to find the appropriate upstream solution.

LorenzMeier commented:
@davids5 Can you review and upstream?

davids5 commented:
@LorenzMeier 

Unfortunately this PR is against the old repo layout. It has commits that span multiple repos. So the changes need to be hand merged across the ntttx and NxWidgets repos.  

@nuttx If you could have a look at **just the deltas** here. (There are others changes that will not be submitted to you in the files) and see if you are OK with the changes. 

If you are I will do I will do separate PR on BB on the ntttx and NxWidgets repos.  

Especially : https://github.com/PX4/NuttX/pull/85/commits/8256188d956ca0e163f1948ce4c4052e0f266c7a#diff-411903c5faf2f77ddc82bdb94ee83c76R65 

It is not near to the coding standard. So let me know if you want all those wrapped with proper title blocks.

CarloWood commented:
@davids5 Can you otherwise already commit this to our repository (old layout), so I can start using it and can continue on master with what I'm doing? Especially the cctype changes indeed, the other commit is just a courtesy towards Greg, who wrote me in email, when I brought the cctype problem up, that C++ wasn't really his thing (and the cctype logic was simply brought in from PX4 to begin with). I could live with a delay on the first commit (removal of cunistd et al) although, as someone who lives and breathes C++, I hope that will make it sooner than later too ;).

Please let me know if I can do something, like applying this on the nuttx_v3 branch, it would be no problem at all.

davids5 commented:
@CarloWood 

To move this forward, this is what I would suggest: 

Please breakout 8256188 into a separate PR - follow the NuttX coding standard

 http://nuttx.org/doku.php?id=documentation:codingstandard

It would be helpful for you to do the PR's against nuttx_v3 but getting this upstream requires we 
verify that every graphics and NxWM and C++ configuration still builds in stand alone Nuttx. 

There will require identifying the affected configs and building them. There is a tool here: https://beta.groups.yahoo.com/neo/groups/nuttx/files see test-files.zip  for that.

CarloWood commented:
@davids5 What coding style are you referring to in particular? When I look through the source code, I find many different styles. There is nothing in the coding standard that talks about inline functions inside a namespace (it is clearly written for C anyway, not C++) so it's unclear if I should use the /* namespace */ comment directly above the 'namespace' statement, or the /* inline */ comment directly above the 'inline' declarations, or both. I looked for examples in the code that are at least inline functions, but they use various styles (ie, tabs as indentation and just one-liner comments to describe the functions - not a complete documentation as is "required" by the coding standard, which in deed would be extremely silly for THESE functions, which are exactly the same as their counter part macro's in ctype.h and well documented in the C and C++ documentation of the standard functions in question ;).

The thing that did strike me though is that every inline function in NuttX is marked static. My feeling was that one should *seldom* need to use 'static' in C++, if at all (it was even deprecated for a while), and using it for inline functions has a risky result that the code might grow, because essentially you tell the compiler to generate a new function for each compilation unit (that includes the header). Now, IF the function is indeed inlined everywhere (but you can't know that, it's the compiler who decides if it should) then that code would not be generated and the function disappears from the compilation unit; but that won't be the case for all compilation units imho, so the over all code will grow compared to using just 'inline', getting a single instantiation of the function somewhere (that is, the compiler will generate an instantiation in a special sector in all compilation units and the linker will then only use one of them, just like it does with templates). To make a long story short: personally I'd not use 'static', if you want me to then please tell me.

LorenzMeier commented:
Can we close this now?

dagar committed:
[PX4: REJECTED] stm32f7 dma capable hack for SPI DMA usage

 - remove incomplete CONFIG_STM32F7_SPI_DMA_THRESHOLD
 - fix missing DMAMAP_SPI2 alternates (verify)

dagar commented:
TODO: address the `arch/arm/src/stm32f7/hardware/stm32f76xx77xx_dma.h` and `arch/arm/src/stm32f7/stm32_spi.c` issues upstream.

Another potential solution to the larger problem (dma capable buffer checks + dcache) would be to extend the NuttX SPI_DMATHRESHOLD feature some architectures have to stm32, stm32f7, stm32h7, etc and then disable `CONFIG_STM32F7_DMACAPABLE`. The threshold would be set to capture all SPI transfers other than IMU drivers emptying FIFOs. In those cases we'd be responsible for supplying an appropriate DMA capable buffer.

https://github.com/PX4/NuttX/compare/pr-stm32_SPI_DMATHRESHOLD?expand=1
